Objects/unicodeobject.c

```diff
@@ -2272,34 +2269,74 @@ PyUnicode_FromString(const char *u)
     return PyUnicode_DecodeUTF8Stateful(u, (Py_ssize_t)size, NULL, NULL);
 }
 
+
 PyObject *
 _PyUnicode_FromId(_Py_Identifier *id)
 {
-    if (!id->object) {
-        id->object = PyUnicode_DecodeUTF8Stateful(id->string,
-                                                  strlen(id->string),
-                                                  NULL, NULL);
-        if (!id->object)
+    PyInterpreterState *interp = _PyInterpreterState_GET();
+    struct _Py_unicode_ids *ids = &interp->unicode.ids;
+
+    if (id->index < 0) {
+        struct _Py_unicode_runtime_ids *rt_ids = &interp->runtime->unicode_ids;
+
+        PyThread_acquire_lock(rt_ids->lock, WAIT_LOCK);
+        id->index = rt_ids->next_index;
+        rt_ids->next_index++;
+        PyThread_release_lock(rt_ids->lock);
+
+        assert(id->index >= 0);
+    }
+
+    PyObject *object;
+    if (id->index < ids->size) {
```

Comment:
It may be worth to save `id->index` as a local variable. Especially if make it volatile.

Reply:
Done.